File: backend/CoverMe.Backend.Core/Helpers/ProcessHelper.cs
using System.Diagnostics;
using CoverMe.Backend.Core.Extensions;
using CoverMe.Backend.Core.Helpers.Abstractions;
using CoverMe.Backend.Core.Models.Process;
using Microsoft.Extensions.Logging;

namespace CoverMe.Backend.Core.Helpers;

public class ProcessHelper : IProcessHelper
{
    #region Members

    private readonly ILogger<ProcessHelper> _logger;
    private bool _isDotCoverCliInstalled;
    private bool _isReportGeneratorInstalled;
    private static readonly SemaphoreSlim DotCoverInstallationLock = new(1, 1);
    private static readonly SemaphoreSlim ReportGeneratorInstallationLock = new(1, 1);

    #endregion

    #region Constructors

    public ProcessHelper(ILogger<ProcessHelper> logger)
    {
        _logger = logger;
    }

    #endregion

    #region Public methods

    public async Task<ProcessResponse> DotCoverCli(DotCoverCliOptions options)
    {
        if (!_isDotCoverCliInstalled)
        {
            if (!await EnsureDotCoverCliInstalled())
            {
                return new ProcessResponse
                {
                    ExitCode = 1,
                    Error = "Failed to install dotcover cli",
                    Output = string.Empty
                };
            }
        }

        var arguments = new List<string>
        {
            "dotcover",
            options.Command.Description(),
            $"--ReportType={options.ReportType.Description()}",
            $"--Output=\"{options.OutputPath}\"",
        };

        if (options.HideAutoProperties)
        {
            arguments.Add("--HideAutoProperties");
        }

        arguments.AddRange([
            "--",
            "test"
        ]);

        if (options.NoBuild)
        {
            arguments.Add("--no-build");
        }

        arguments.Add($"\"{options.ProjectFolderPath}\"");

        return await Execute(
            "dotnet",
            arguments,
            options.ProjectFolderPath
        );
    }

    public async Task<bool> EnsureDotCoverCliInstalled()
    {
        await DotCoverInstallationLock.WaitAsync();

        try
        {
            if (_isDotCoverCliInstalled) return true;

            var response = await Execute(
                "dotnet",
                [
                    "tool",
                    "install",
                    "--global",
                    "JetBrains.dotCover.CommandLineTools"
                ],
                "."
            );

            _isDotCoverCliInstalled = response.ExitCode == 0;

            if (!_isDotCoverCliInstalled)
            {
                _logger.LogWarning(
                    "Failed to install dotCover CLI: ExitCode={ExitCode}, Output={Output}, Error={Error}",
                    response.ExitCode, response.Output, response.Error);
            }
            else
            {
                _logger.LogTrace("DotCover CLI: {Output}", response.Output);
            }

            return _isDotCoverCliInstalled;
        }
        finally
        {
            DotCoverInstallationLock.Release();
        }
    }

    public async Task<ProcessResponse> ReportGenerator(ReportGeneratorOptions options)
    {
        if (!_isReportGeneratorInstalled)
        {
            if (!await EnsureReportGeneratorInstalled())
            {
                return new ProcessResponse
                {
                    ExitCode = 1,
                    Error = "Failed to install report generator",
                    Output = string.Empty
                };
            }
        }

        var arguments = new List<string>
        {
            $"-reports:\"{options.ReportFilePath}\"",
            $"-targetdir:\"{options.OutputFolderPath}\"",
        };

        return await Execute(
            "reportgenerator",
            arguments,
            Path.GetDirectoryName(options.ReportFilePath) ?? "."
        );
    }

    public async Task<bool> EnsureReportGeneratorInstalled()
    {
        await ReportGeneratorInstallationLock.WaitAsync();

        try
        {
            if (_isReportGeneratorInstalled) return true;

            var response = await Execute(
                "dotnet",
                [
                    "tool",
                    "install",
                    "--global",
                    "dotnet-reportgenerator-globaltool"
                ],
                "."
            );

            _isReportGeneratorInstalled = response.ExitCode == 0;

            if (!_isReportGeneratorInstalled)
            {
                _logger.LogWarning(
                    "Failed to install report generator: ExitCode={ExitCode}, Output={Output}, Error={Error}",
                    response.ExitCode, response.Output, response.Error);
            }
            else
            {
                _logger.LogTrace("Report Generator: {Output}", response.Output);
            }

            return _isReportGeneratorInstalled;
        }
        finally
        {
            ReportGeneratorInstallationLock.Release();
        }
    }

    public async Task<ProcessResponse> Execute(
        string command,
        IEnumerable<string> arguments,
        string workingDirectory
    )
    {
        var argumentsStr = string.Join(" ", arguments);

        _logger.LogInformation(
            "ProcessHelper: Executed: Command={Command}; Arguments={Arguments}; WorkingDirectory={WorkingDirectory};",
            command,
            argumentsStr,
            workingDirectory
        );
        var process = new Process
        {
            StartInfo =
            {
                FileName = command,
                Arguments = argumentsStr,
                WorkingDirectory = workingDirectory,
                UseShellExecute = false,
                CreateNoWindow = true,
                RedirectStandardError = true,
                RedirectStandardOutput = true
            },
        };

        if (!process.Start()) throw new ApplicationException();

        var output = await process.StandardOutput.ReadToEndAsync();
        var error = await process.StandardError.ReadToEndAsync();
        await process.WaitForExitAsync();

        _logger.LogInformation(
            "ProcessHelper: Executed: Command={Command}; Arguments={Arguments}; WorkingDirectory={WorkingDirectory}; ExitCode={ExitCode}; Output={Output}; Error={Error};",
            command,
            argumentsStr,
            workingDirectory,
            process.ExitCode,
            output,
            error
        );

        return new ProcessResponse
        {
            ExitCode = process.ExitCode,
            Output = output,
            Error = error
        };
    }

    #endregion
}

File: src/src/rider/main/kotlin/com/jetbrains/rider/plugins/coverme/helpers/ProcessHelper.kt
package com.jetbrains.rider.plugins.coverme.helpers

import com.intellij.openapi.components.Service
import com.jetbrains.rdclient.util.idea.toIOFile
import com.jetbrains.rider.plugins.coverme.enums.process.DotCoverCliCommand
import com.jetbrains.rider.plugins.coverme.enums.process.DotCoverCliReportType
import com.jetbrains.rider.plugins.coverme.models.process.DotCoverCliOptions
import com.jetbrains.rider.plugins.coverme.models.process.ProcessResponse
import com.jetbrains.rider.plugins.coverme.models.process.ReportGeneratorOptions
import kotlinx.coroutines.sync.Semaphore
import java.io.File

@Service(Service.Level.PROJECT)
class ProcessHelper {
    private var _isDotCoverCliInstalled: Boolean = false
    private var _isReportGeneratorInstalled: Boolean = false
    private val _dotCoverCliInstallationLock = Semaphore(1)
    private val _reportGeneratorInstallationLock = Semaphore(1)

    fun dotCoverCli(options: DotCoverCliOptions): ProcessResponse {
        if (!_isDotCoverCliInstalled) {
            if (!ensureDotCoverCliInstalled()) {
                return ProcessResponse(1, "", "Failed to install dotCover CLI")
            }
        }

        val arguments = mutableListOf(
            "dotcover",
            when (options.command) {
                DotCoverCliCommand.COVER_DOTNET -> "cover-dotnet"
                else -> throw UnsupportedOperationException()
            },
            "--ReportType=${
                when (options.reportType) {
                    DotCoverCliReportType.HTML -> "HTML"
                    DotCoverCliReportType.DETAILED_XML -> "DetailedXML"
                    else -> throw UnsupportedOperationException()
                }
            }",
            "--Output=${options.outputPath}"
        )

        if (options.hideAutoProperties) {
            arguments.add("--HideAutoProperties")
        }

        arguments.add("--")
        arguments.add("dotnet")
        arguments.add("test")

        if (options.noBuild) {
            arguments.add("--no-build")
        }

        arguments.add("\"${options.projectFolderPath}\"")

        return execute(
            "dotnet",
            arguments.toTypedArray(),
            options.projectFolderPath
        )
    }

    fun reportGenerator(options: ReportGeneratorOptions): ProcessResponse {
        if (!_isReportGeneratorInstalled) {
            if (!ensureReportGeneratorInstalled()) {
                return ProcessResponse(1, "", "Failed to install report generator")
            }
        }

        val arguments = mutableListOf(
            "-reports:\"\"${options.reportFilePath}\"\"",
            "-targetdir:\"\"${options.outputFolderPath}\"\"",
        )

        return execute(
            "reportgenerator",
            arguments.toTypedArray(),
            File(options.reportFilePath).parent
        )
    }

    private fun ensureReportGeneratorInstalled(): Boolean {
        if (!_reportGeneratorInstallationLock.tryAcquire()) return false

        try {
            if (_isReportGeneratorInstalled) return true

            val response = execute(
                "dotnet",
                arrayOf(
                    "tool",
                    "install",
                    "--global",
                    "dotnet-reportgenerator-globaltool"
                ),
                "."
            )

            _isReportGeneratorInstalled = response.exitCode == 0
            return _isReportGeneratorInstalled
        } finally {
            _reportGeneratorInstallationLock.release()
        }
    }

    private fun ensureDotCoverCliInstalled(): Boolean {
        if (!_dotCoverCliInstallationLock.tryAcquire()) return false

        try {
            if (_isDotCoverCliInstalled) return true

            val response = execute(
                "dotnet",
                arrayOf(
                    "tool",
                    "install",
                    "--global",
                    "JetBrains.dotCover.CommandLineTools"
                ),
                "."
            )

            _isDotCoverCliInstalled = response.exitCode == 0
            return _isDotCoverCliInstalled
        } finally {
            _dotCoverCliInstallationLock.release()
        }
    }

    private fun execute(command: String, arguments: Array<String>, workingDirectory: String): ProcessResponse {
        val process = ProcessBuilder(command, *arguments).directory(workingDirectory.toIOFile())
            .redirectInput(ProcessBuilder.Redirect.INHERIT)
            .redirectError(ProcessBuilder.Redirect.INHERIT)
            .start()

        val outputLines: MutableList<String> = mutableListOf()
        process.inputStream.bufferedReader()
            .use {
                outputLines.addAll(it.readLines())
            }

        val errorLines: MutableList<String> = mutableListOf()
        process.errorStream.bufferedReader()
            .use {
                errorLines.addAll(it.readLines())
            }

        return ProcessResponse(
            process.exitValue(), outputLines.joinToString(), errorLines.joinToString()
        )
    }
}